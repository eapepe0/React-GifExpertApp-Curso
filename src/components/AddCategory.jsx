import React from 'react'
import { useState } from 'react'

export const AddCategory = ({onNewCategory}) => {
    const [inputValue, setInputValue] = useState('')

    const onInputChange = ({target}) =>{
        setInputValue(target.value)
    }

    const onSubmit = (event) =>{
    
        event.preventDefault();

        if(inputValue.trim().length <=1) return;
        onNewCategory(inputValue.trim())
        setInputValue('')

    }
  return (
    <form onSubmit={onSubmit}>
        <input 
            type="text" 
            name="" id="" 
            placeholder='Buscar gifs' 
            value= {inputValue}
            onChange= { onInputChange }
        />
    </form>
  )
}


/**
 *  Recibimos una funcion del padre de AddCategory (GifExpert) , llamada onAddCategory
 * 
 *  en el <form> , tenemos un evento onSubmit que se llama cuando se apreta <Enter> en el <input>
 *  llama a la funcion onSubmit
 *  en el elemento html <input> , podemos ver que el value es igual al inputValue que es una const estado
 *  y cada vez que cambia llama a la funcion onInputChange (linea 7), que recibe un event, del cual extraemos el target
 *  que es el elemento html que lo dispara y el value , el valor que escribimos en el input
 *  le agregamos a inputValue el valor escrito en el input mediante (setInputValue) (linea 8)
 * 
 *  la funcion onSubmit recibe como valor un evento , para prevenir que al apretar <Enter> se actualice la pagina
 *  usamos event.preventDefault(); (linea 13)
 *  nos preguntamos si lo ingresa sin espacios (trim) es menor o igual a 1 no hacemos nada (return) (linea 15)
 *  sino llamamos a la funcion onNewCategory que en realidad es un puntero que apunta a la funcion onAddCategory del padre 
 *  (ver onAddCategory de GifExpertApp) , que lo que hace es ingresar a la constante estado categories lo ingresado
 *  por el input , despues de ingresar borramos el valor del <input>
 *  
 */