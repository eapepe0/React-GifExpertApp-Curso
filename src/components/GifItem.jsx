import React from 'react'
import PropTypes from 'prop-types'

export const GifItem = ({title, url , id}) => {
  return (
    <div className="card">
        <img src={url} alt={title} />
        <p>{title}</p>
    </div>
  )
}




GifItem.propTypes={
  title : PropTypes.string.isRequired,
  url : PropTypes.string.isRequired,
  id : PropTypes.number
}
/**
|--------------------------------------------------
|  este componente dibuja la tarjeta donde se muestra
|  la imagen con el titulo , recibe los valores
| |--------------------------------------------------
*/

/**
|--------------------------------------------------
| 1 . AÃ±adir proptypes
|   a. titulo obligatorio
|   b. url obligatorio
|
| 2. Evaluar snapshot 
|--------------------------------------------------
*/