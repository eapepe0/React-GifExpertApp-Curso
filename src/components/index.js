export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem';

/**
|--------------------------------------------------
| Esto es llamado archivo barril
| Podemos exportar mas facilmente los archivos
|
| import GifGrid from './components/GrifGrid'
| import AddCategory from './components/AddCategory'
|
|   a esto 
|
|  import { AddCategory , GifGrid} from './components';
|--------------------------------------------------
*/