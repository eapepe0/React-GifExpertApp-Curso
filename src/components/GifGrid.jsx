import React, { useEffect , useState } from 'react'
import { GifItem } from './GifItem';
import { useFetchGifs } from '../hooks/useFetchGifs';
import PropTypes from 'prop-types'

export const GifGrid = ({category}) => {
    const {images , isLoading} = useFetchGifs(category)
    return (
        <>
            <h3>{category}</h3>
            {
                isLoading && (<h2>Cargando...</h2>)
            }
            <div className='card-grid'>
                {
                    images.map((imagen) =>(
                        <GifItem key={imagen.id} {...imagen}/>
                    ))
                }

            </div>
        </>
    )
}





GifGrid.propTypes={
   category : PropTypes.string.isRequired  ,
}

/**
 * por cada categoria (busqueda)
 * llamamos la funcion useFetchGifs con el valor a buscar
 * nos devuelve un array con objetos con los valores id , titulo ,url
 * 
 * despues mapeamos por cada item en el array images
 * un componente que tome el id y despues destructuramos los valores , 
 * como ya el objeto devuelto tiene su title , su id y su url
 * los usamos en el componente GifItem
 */


/**
|--------------------------------------------------
| cada vez que se hace un cambio se vuelve a renderizar 
| el componente , ciertas funciones vamos a querer que 
| se ejecuten una sola vez , como por ej una llamada a
| un api, 
|     
|--------------------------------------------------
*/

/**
|--------------------------------------------------
| useEffect debe devolver una funcion , no una promesa
|
| por eso no podemos usar:
|
|
| 
| const [images, setImages] = useState([]);
| 
| useEffect( async () =>{
|    const newImages = await getGifs(category);
|    setImages(newImages)
| },[])
|
|
|--------------------------------------------------
*/

/**
|--------------------------------------------------
| podemos usar :
|
| useEffect( async () =>{
|    getGifs(category);
|         .then(newImages => setImages(newImages))
|        // cuando la funcion devuelve la promesa
|        // newImages es donde hace el return
|        // getGifs
|        // entonces newImages es igual setImages
| },[])
|
|
|--------------------------------------------------
*/

/**
 * podemos llamar al GifItem de esta forma pasando las props asi
 * <GifItem key={imagen.id} title= {imagen.url} url = {imagen.url}/>
 * 
 * o usando un spread y recibiendo automanticamente
 * 
 * <GifItem key={imagen.id} {...imagen}/>
 */

/**
 * first
 */

/**
|--------------------------------------------------
| en la linea 12, hace un renderizado condicional
| si la const isLoading es verdadera renderiza
| de lo contrario no
|--------------------------------------------------
*/