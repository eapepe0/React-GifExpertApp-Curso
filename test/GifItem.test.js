import {render , screen} from '@testing-library/react'
import {GifItem} from '../src/components/GifItem'
import React from 'react'

describe('Pruebas en GridItem',()=>{

    const title = 'Yo soy el titulo';
    const url = 'https://one-punch.com/saitama.jpg';
    const id = 11;

    test('debe hacer match con el snapshot',()=>{

        const { container } = render(<GifItem title = {title} url = {url} id = {id} />);
        expect( container ).toMatchSnapshot();
    }) 
    test('debe mostrar la imagen con el URL y el ALT indicado', () => {
        render(<GifItem title = {title} url = {url} id = {id} />);

        expect(screen.getByRole( 'img' ).src).toBe( url ) // sacamos el src de la imagen y lo comparamos con la url
        expect(screen.getByRole( 'img' ).alt).toBe( title ) // sacamos el alt de la imagen y lo comparamos con el title
        
        const {src , alt} = screen.getByRole('img'); // desestructuramos src y alt de la imagen

        expect(src).toBe(url)
        expect(alt).toBe(title)
    });

    test('debe mostrar el titulo', () => {
        render(<GifItem title = {title} url = {url} id = {id} />); // renderizamos el componente
        expect(screen.getByText(title)).toBeTruthy( ) // esperamos que el titulo buscado sea verdadero
    });
});