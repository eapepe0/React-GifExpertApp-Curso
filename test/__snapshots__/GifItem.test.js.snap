// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en GridItem debe hacer match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Yo soy el titulo"
      src="https://one-punch.com/saitama.jpg"
    />
    <p>
      Yo soy el titulo
    </p>
  </div>
</div>
`;
